app/src/ui/diff/text-diff-expansion.ts

```diff
@@ -0,0 +1,402 @@
+import {
+  DiffHunk,
+  DiffHunkHeader,
+  DiffLine,
+  DiffLineType,
+  ITextDiff,
+} from '../../models/diff'
+
+export const DiffExpansionDistance = 20
+
+export type ExpansionKind = 'up' | 'down'
+
+/** Interface to represent the expansion capabilities of a hunk header. */
+interface IHunkHeaderExpansionInfo {
+  /** True if the hunk header can be expanded down. */
+  readonly isExpandableDown: boolean
+
+  /** True if the hunk header can be expanded up. */
+  readonly isExpandableUp: boolean
+
+  /** True if the hunk header can be expanded both up and down. */
+  readonly isExpandableBoth: boolean
+
+  /**
+   * True if the hunk header represents a short gap that, when expanded, will
+   * result in merging this hunk and the hunk above.
+   */
+  readonly isExpandableShort: boolean
+}
+
+/** Builds the diff text string given a list of hunks. */
+function getDiffTextFromHunks(hunks: ReadonlyArray<DiffHunk>) {
+  // Grab all hunk lines and rebuild the diff text from it
+  const newDiffLines = hunks.reduce<ReadonlyArray<DiffLine>>(
+    (result, hunk) => result.concat(hunk.lines),
+    []
+  )
+
+  return newDiffLines.map(diffLine => diffLine.text).join('\n')
+}
+
+/** Merges two hunks into one. */
+function mergeDiffHunks(hunk1: DiffHunk, hunk2: DiffHunk): DiffHunk {
+  const firstHunk1Line = hunk1.lines[0]
+  const allHunk1LinesButFirst = hunk1.lines.slice(1)
+  const allHunk2LinesButFirst = hunk2.lines.slice(1)
```

Comment:
This removes the first line in both hunks, because those are the `@@ -1,32 +1,34 @@` header lines that will be replaced by a new one for the resulting hunk.